lib/plugins/awsCompileFunctionsToResources/awsCompileFunctionsToResources.js

```diff
@@ -0,0 +1,111 @@
+'use strict';
+
+const BbPromise = require('bluebird');
+const forEach = require('lodash').forEach;
+const merge = require('lodash').merge;
+const path = require('path');
+
+class AwsCompileFunctionsToResources {
+  constructor(serverless) {
+    this.serverless = serverless;
+
+    // default properties
+    this.functions = []; // extracted functions from the serverless.yaml file as JS objects
+    this.functionResources = []; // AWS resource definitions for the functions
+    this.cloudFormationResult = {};
+
+    // TODO: remove because it relies on awsResourcesDeploy hooks
+    this.commands = {
+      compile: {
+        commands: {
+          functions: {
+            usage: 'Compiles the function definition in serverless.yaml to a CloudFormation stack',
+            lifecycleEvents: [
+              'compile',
+            ],
+          },
+        },
+      },
+    };
+
+    this.hooks = {
+      'compile:functions:compile': (options) => {
+        this.options = options;
+        return BbPromise.bind(this)
+          .then(this.extractFunctions)
+          .then(this.createFunctionResources)
+          .then(this.addFunctionResourcesToCFMainTemplate)
+          .catch((exception) => {
+            throw new this.serverless.classes.Error(exception);
+          });
+      },
+    };
+  }
+
+  extractFunctions() {
+    const functions = this.serverless.service.functions;
+
+    forEach(functions, (value, key) => {
+      // check if it's the function and not the name_template property
+      if (key !== 'name_template') {
+        const functionObject = {
+          [key]: value,
+        };
+
+        this.functions.push(functionObject);
+      }
+    });
+    return BbPromise.resolve();
+  }
+
+  createFunctionResources() {
+    const functionTemplate = this.serverless.utils.readFileSync(
+      path.join(__dirname, 'templates', 'cf-function-template.json'));
```

Comment:
we really don't need that template file. You can easily create that template object here in code and hold it in memory. Most of the properties in the template will be overwritten in memory anyway like you're doing below.